Add interface definitions for auto fix support

wadimw opened the pull request.

wadimw committed:
Add typechecking support

wadimw committed:
Add property definitions for the typechecking

wadimw committed:
Fix unsupported type syntax

wadimw committed:
Add interface definitions for autofixing
- add and export definitions for Fix and Fixer
- include Fix in Result
- include Fixers in Plugin
- add interface for Parser writing back to file

wadimw commented:
@ehoogerbeets I think it's easier to read in a per-commit manner

wadimw commented:
Regarding the HTML tags replaced with Markdown syntax in JSDoc, I still left those changes in here because VSCode completely leaves out the parts enclosed in HTML (like lists) so huge chunks of text just didn't render for me at all. I know it's probably uglier in HTML render but should still be readable there. Overall, we need to get this sorted out (and as I said before I think it should be easier to switch to JSDoc with the [builtin Markdown plugin](https://jsdoc.app/plugins-markdown.html); note this is a different issue than the Typescript-specific syntax and given it's built in, it shouldn't be buggy)

ehoogerbeets commented on src/Fix.js:
````diff
@@ -0,0 +1,120 @@
+ * If there are no Fix and Fixer with type `string`,
+ * they could be implemented like in the following example:
+ * ```ts
+ * class StringFix extends Fix {
````
Nothing else in the documentation is in typescript, and the code itself is not typescript. I would prefer converting it back to JS.

wadimw replied:
Changed in 08090d6

ehoogerbeets commented on src/Fix.js:
```diff
@@ -0,0 +1,120 @@
+     */
+    applied = false;
+}
+
```
Do we not want to formalize that Fixes contain one or more FixCommands?

wadimw replied:
I was thinking about it but decided against it in the end, because the Fix class is already an abstract container for any data that a given (matching) fixer might need to perform a fix - so there is no point in introducing an additional abstract container class FixCommand which has literally no properties of its own.

You define additional properties of a Fix when declaring a `class SpecificFix extends Fix`, so then within the fixer.apply You will know about them anyway, since You expect that the fixer.fix receives only instances of SpecificFix.

ehoogerbeets commented on src/Parser.js:
```diff
@@ -17,54 +17,101 @@
+import NotImplementedError from "./NotImplementedError.js";
+
 /**
- * @class common SPI for parser plugins
```
SPI - Service Provider Interface - when the calling code creates an interface to document what they expect of your code. ie. docs for plugins - https://en.wikipedia.org/wiki/Service_provider_interface

API - Applications Provider Interface - when you create an interface to document what someone else should expect of your code

wadimw replied:
Reverted in 4dda910

ehoogerbeets committed:
Fixes to the jsdocs

- added plugin in jsdoc.json so we can use markdown syntax in the
  the jsdoc comments
- converts to markdown in docs/i18nlint-common.md
- converts to HTML in the docs/index.html

wadimw committed:
Revert typo fix (was not a typo)

wadimw committed:
Change example code in Fix docs to JS; refresh doc

wadimw committed:
Add release notes

wadimw committed:
Bump version for publish
